File: 203.移除链表元素.js
/*
 * @lc app=leetcode.cn id=203 lang=javascript
 *
 * [203] 移除链表元素
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} val
 * @return {ListNode}
 */
// 使用哨兵节点
var removeElements = function (head, val) {
    let sentinel = new ListNode(null)
    sentinel.next = head
    let prev = sentinel
    let cur = head
    while(cur) {
        if (cur.val === val) {
            let next = prev.next = cur.next
            cur.next = null
            cur = next
        } else {
            prev = cur
            cur = cur.next
        }
    }
    return sentinel.next
}

// 判断prev节点是否存在
var removeElements1 = function (head, val) {
    let cur = head
    let prev
    while (cur) {
        if (cur.val === val) {
            if (prev) {
                prev.next = cur.next
            } else {
                head = cur.next
            }
            let next = cur.next
            cur.next = null
            cur = next
        } else {
            prev = cur
            cur = cur.next
        }
    }
    return head
};
// @lc code=end

File: 203.移除链表元素.java
/*
 * @lc app=leetcode.cn id=203 lang=java
 *
 * [203] 移除链表元素
 */

// @lc code=start
/**
 * Definition for singly-linked list. public class ListNode { int val; ListNode
 * next; ListNode(int x) { val = x; } }
 */
// public class ListNode {
//     int val;
//     ListNode next;

//     ListNode(int x) {
//         val = x;
//     }
// }

class Solution {
    public ListNode removeElements(ListNode head, int val) {
        ListNode sentinel = new ListNode(0);
        sentinel.next = head;
        ListNode prev = sentinel, cur = head;
        while (cur != null) {
            if (cur.val == val) {
                prev.next = cur.next;
                cur.next = null;
            } else {
                prev = cur;
            }
            cur = prev.next;
        }
        return sentinel.next;
    }
}
// @lc code=end
